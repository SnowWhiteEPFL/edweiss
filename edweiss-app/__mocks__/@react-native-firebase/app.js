export default {
    initializeApp: jest.fn(),
    auth: jest.fn(),
    firestore: jest.fn(),
};
