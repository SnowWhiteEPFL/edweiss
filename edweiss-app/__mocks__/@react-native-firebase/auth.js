export default {
    onAuthStateChanged: jest.fn(),
    signInWithEmailAndPassword: jest.fn(),
    signOut: jest.fn(),
};
