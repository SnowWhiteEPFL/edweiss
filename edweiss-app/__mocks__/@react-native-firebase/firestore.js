const firestore = {
    collection: jest.fn(() => ({
        doc: jest.fn(() => ({
            onSnapshot: jest.fn(),
        })),
    })),
    doc: jest.fn(),
    onSnapshot: jest.fn(),
};

export default firestore;
