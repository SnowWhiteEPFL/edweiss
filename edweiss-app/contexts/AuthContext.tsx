import React, { ReactNode, useContext, useState } from 'react';

export interface AppUser {
	username: string,
	age: number,
	sciper: string
}

const AuthContext = React.createContext<AppUser>({} as unknown as AppUser);

export function useAuth() {
	return useContext(AuthContext);
}

export function AuthContextProvider(props: { children: ReactNode }) {
	const [user, setUser] = useState<AppUser>({
		username: "user",
		age: 20,
		sciper: "123456"
	});

	return (
		<AuthContext.Provider value={user}>
			{props.children}
		</AuthContext.Provider>
	)
}
