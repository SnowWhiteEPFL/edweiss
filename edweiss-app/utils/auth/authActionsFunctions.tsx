import { callFunction } from '@/config/firebase';
import { Auth } from '@/model/users';
import Toast from 'react-native-toast-message';

export async function switchPermissionsAction() {
    const res = await callFunction(Auth.Functions.switchPermissions, {});
    if (res.status != 1) {
        console.error(res.error);
        Toast.show({
            type: 'error',
            text1: "Permission switch failed",
        });
        return { ...res, error: new Error(res.error) };
    }
    Toast.show({
        type: 'success',
        text1: "You have switched permissions",
    });
    return res;
}

