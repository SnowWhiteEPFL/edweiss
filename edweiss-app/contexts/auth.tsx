import auth, { FirebaseAuthTypes } from '@react-native-firebase/auth';
import React, { useContext, useEffect, useState } from 'react';

type User = FirebaseAuthTypes.User;

interface AuthInterface {
	userLoggedIn: boolean,
	authUser: User,
	isLoading: boolean
}

const AuthContext = React.createContext<AuthInterface>(null as unknown as AuthInterface);

export function useAuth() {
	return useContext(AuthContext);
}

export function SessionProvider(props: React.PropsWithChildren) {
	const [currentUser, setCurrentUser] = useState<User>();
	const [userLoggedIn, setUserLoggedIn] = useState(false);
	const [loading, setLoading] = useState(true);

	useEffect(() => {
		const unsubscribe = auth().onAuthStateChanged(initializeUser);
		return unsubscribe;
	}, []);

	async function initializeUser(user: User | null) {
		console.log("Auth: Initialize user");
		setLoading(true);

		if (user) {
			setCurrentUser(user);
			setUserLoggedIn(true);
		} else {
			setCurrentUser(undefined);
			setUserLoggedIn(false);
		}

		setLoading(false);
	}

	const value: AuthInterface = {
		authUser: currentUser as User,
		userLoggedIn,
		isLoading: currentUser == undefined ? loading : false
	}

	return (
		<AuthContext.Provider value={value}>
			{props.children}
		</AuthContext.Provider>
	);
}
