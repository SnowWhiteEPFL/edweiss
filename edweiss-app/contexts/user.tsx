import { Collections, getDocumentRef } from '@/config/firebase';
import { AppUser } from '@/model/users';
import { onSnapshot } from '@react-native-firebase/firestore';
import React, { useContext, useEffect, useState } from 'react';
import { useAuth } from './auth';

interface UserInterface {
	user: AppUser,
	loaded: boolean
}

const UserContext = React.createContext<UserInterface>({} as UserInterface);

export function useUser() {
	return useContext(UserContext);
}

export function UserProvider(props: React.PropsWithChildren) {
	const [user, setAppUser] = useState<AppUser>();
	const [loaded, setLoaded] = useState(false);

	const { authUser } = useAuth();

	useEffect(() => {
		if (authUser == undefined) {
			setAppUser(undefined);
			setLoaded(true);
			return;
		}

		setLoaded(false);

		const userRef = getDocumentRef(Collections.users, authUser.uid);

		userRef.get().then(user => {
			setAppUser(user.data());
			setLoaded(true);
		});

		const unsubscribe = onSnapshot(userRef, async (doc) => {
			if (doc.exists)
				setAppUser(doc.data());
			else
				setAppUser(undefined);

			setLoaded(true);
		});

		return unsubscribe;
	}, [authUser]);

	const value: UserInterface = {
		user: user as AppUser,
		loaded
	}

	return (
		<UserContext.Provider value={value}>
			{props.children}
		</UserContext.Provider>
	)
}
