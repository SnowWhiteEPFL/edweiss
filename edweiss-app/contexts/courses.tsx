import { CollectionOf, Document } from '@/config/firebase'
import { useDocsWithIds } from '@/hooks/firebase/firestore'
import { Course, CourseID } from '@/model/school/courses'
import React, { useContext } from 'react'
import { useUser } from './user'

type CourseInterface = Document<Course>[] | undefined

const CourseContext = React.createContext<CourseInterface>(undefined)

export function useCourses() {
	return useContext(CourseContext)
}

export function useCourse(id: CourseID): Document<Course> | undefined {
	const courses = useCourses();

	if (courses == undefined)
		return undefined;

	return courses.filter(c => c.id == id)[0];
}

export function CoursesProvider(props: React.PropsWithChildren) {
	const { user } = useUser();

	const courses = useDocsWithIds(CollectionOf<Course>("courses"), user.courses ?? []);

	return (
		<CourseContext.Provider value={courses}>
			{props.children}
		</CourseContext.Provider>
	)
}
