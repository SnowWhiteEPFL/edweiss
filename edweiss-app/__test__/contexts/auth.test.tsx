import { AuthSessionProvider } from '@/contexts/auth';
import { FirebaseAuthTypes } from '@react-native-firebase/auth';
import { render } from '@testing-library/react-native';
import { Text } from 'react-native';

const firebaseUser = { displayName: "John Doe", uid: 'firebase-test-uid', email: 'firebase-test@example.com' };

jest.mock('@react-native-firebase/auth', () => ({
	__esModule: true,
	currentUser: firebaseUser,
	default: jest.fn(() => ({
		onAuthStateChanged: jest.fn((callback: (user: FirebaseAuthTypes.User | null) => void) => {
			callback(firebaseUser as FirebaseAuthTypes.User);
		})
	}))
}));

// const UsingAuthContext: ReactComponent<{}> = () => {
// 	const obj = useAuth();
// 	console.log(obj);
// 	return (<Text>{obj.authUser.displayName}</Text>);
// };

describe("Authentication Context Provider", () => {
	it("should render correctly", () => {
		const Provider = render(<AuthSessionProvider><Text>Some text</Text></AuthSessionProvider>).toJSON();
		expect(Provider).toMatchSnapshot();
	});

	// it("should not useAuth", () => {
	// 	const Component = render(
	// 		<AuthSessionProvider>
	// 			<UsingAuthContext />
	// 		</AuthSessionProvider>
	// 	).toJSON();

	// 	act(() => {
	// 		expect(Component).toMatchSnapshot();
	// 	});
	// });
});
