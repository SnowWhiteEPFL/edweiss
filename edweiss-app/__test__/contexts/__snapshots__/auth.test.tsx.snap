// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Authentication Context Provider should render correctly 1`] = `
<Text>
  Some text
</Text>
`;
